### PR TITLE
[UI] Allow single item deletion from multi-value fields

### LINKED ISSUE #7055: [UI] Multi-value fields should allow to delete single item
image

image

TODO
- Keep `Add ...` button (currently displayed when the list is empty)
- Add a `-` after each item

Examples:
- [ ] Commands/Args in Container form
- [ ] Envrironment variables in Container form (`-` should delete the entire key-value row)
- [ ] Composite commands in Command form
- [ ] Image Build Args

### PR DESCRIPTION
**Pull request by rm3l**
**What type of PR is this:**
/kind feature
/area Web-UI

**What does this PR do / why we need it:**

This adds the ability in the `odo dev` UI to delete single items from multi-value fields:
- container endpoints
- volume mounts
- multi-text fields, like Container Command and Args, or Image Build Args
- multi-key-value fields, like Container Env Vars, Deployment Annotations and Service Annotations
- multi-commands fields, in use in Composite Commands

<details>
<summary>Screenshots</summary>

image

image

image

</details>

**Which issue(s) this PR fixes:**
Fixes #7055 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: ui/src/app/controls/endpoints/endpoints.component.html [hunk 1]
```diff
@@ -1,43 +1,49 @@
...
+                    <mat-option value="ws">ws</mat-option>
+                    <mat-option value="wss">wss</mat-option>
+                    <mat-option value="tcp">tcp</mat-option>
+                    <mat-option value="udp">udp</mat-option>
+                </mat-select>
+            </mat-form-field>
+            <mat-checkbox [attr.data-cy]="'endpoint-secure-'+i" formControlName="secure">Protocol Is Secure</mat-checkbox>
+        </mat-card-content>
+        <mat-card-actions>
+            <button [attr.data-cy]="'endpoint-minus-'+i" mat-icon-button (click)="removeEndpoint(i)">
```
Comment (feloy) [inline ui/src/app/controls/endpoints/endpoints.component.html:41]: You could add a tooltip:

```
matTooltip="delete endpoint"
```

or 

```
[matTooltip]="'delete endpoint '+control.get('name')?.value"
```

File: ui/src/app/controls/multi-command/multi-command.component.html [hunk 2]
```diff
@@ -3,13 +3,15 @@ <h3>{{title}}</h3>
...
         </mat-form-field>
+        <button [attr.data-cy]="'command-minus-'+i"  mat-icon-button (click)="removeCommand(i)">
+            <mat-icon class="tab-icon material-icons-outlined">remove</mat-icon>
+        </button>
     </span>
-    <button *ngIf="form.controls.length > 0" mat-icon-button (click)="addCommand('')">
-        <mat-icon class="tab-icon material-icons-outlined">add</mat-icon>
-    </button>                
-    <button *ngIf="form.controls.length == 0" mat-flat-button (click)="addCommand('')">{{addLabel}}</button>
+    <div>
```
Comment (feloy) [inline ui/src/app/controls/multi-command/multi-command.component.html:14]: I don't know exactly why, the "Add command" button is not displayed at the same place:
- above when editing a container with existing commands
command1

- below when creating a new container, when adding commands
command2

File: ui/src/app/controls/multi-key-value/multi-key-value.component.html [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
                     <mat-label><span>Name</span></mat-label>
                     <input [attr.data-cy]="dataCyPrefix+'-name-'+i" matInput formControlName="name">
                 </mat-form-field>
-                <mat-form-field class="mid-width" appearance="outline">
+                <mat-form-field class="kv-width" appearance="outline">
                     <mat-label><span>Value</span></mat-label>
                     <input [attr.data-cy]="dataCyPrefix+'-value-'+i" matInput formControlName="value">
                 </mat-form-field>
+                <button [attr.data-cy]="dataCyPrefix+'-minus-'+i" mat-icon-button (click)="removeEntry(i)">
+                    <mat-icon class="tab-icon material-icons-outlined">remove</mat-icon>
```
Comment (feloy) [inline ui/src/app/controls/multi-key-value/multi-key-value.component.html:13]: I wonder if a trash icon would be more explicit than a minus sign. Did you try?

File: ui/src/app/controls/multi-command/multi-command.component.html (hunk 2 above)
Comment (rm3l) [reply ui/src/app/controls/multi-command/multi-command.component.html:14]: 🤔 Indeed, good catch. That's weird. I'll take a look.

File: ui/src/app/controls/multi-key-value/multi-key-value.component.html (hunk 3 above)
Comment (rm3l) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html:13]: Haven't tried a trash icon, but it seems to be interesting too:

image

I just tried a small variant of this icon, and it looks even more explicit to me:

image

Thoughts?

Comment (feloy) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html:13]: Yes, the second one seems the best one.

Could you also add some margin on the right of the icon (or remove space on the left), so it is closer to the element it is attached to?

You may be able to "move" the icon with such css:
```
position: relative;
left: 4px;
bottom: 4px; // this one to align the icon vertically with the input
```

File: ui/src/app/controls/multi-command/multi-command.component.html (hunk 2 above)
Comment (rm3l) [reply ui/src/app/controls/multi-command/multi-command.component.html:14]: Ah, I think the issue is that there were no `command` in the Devfile, only `args` set to `tail -f /dev/null`. At least, that's what I saw in the `go` and `nodejs` Devfiles coming from the default registry:

```yaml
- container:
    args:
    - tail
    - -f
    - /dev/null
    endpoints:
    - name: http-go
      targetPort: 8080
    image: registry.access.redhat.com/ubi9/go-toolset:1.18.10-4
    memoryLimit: 1024Mi
    mountSources: true
  name: runtime
```

To make it more explicit, I've just tried to surround each type in a dedicated card, like so:
image

image


Hopefully, it makes things clearer.. Let me know what you think.

File: ui/src/app/controls/multi-key-value/multi-key-value.component.html (hunk 3 above)
Comment (rm3l) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html]: I didn't notice any noteworthy changes in the positioning of this button, but you can take a look at [`4bf895f` (#7084)](https://github.com/redhat-developer/odo/pull/7084/commits/4bf895f272bb23763471ea18affb8ca93c6e6d55)

File: ui/src/app/controls/endpoints/endpoints.component.html (hunk 1 above)
Comment (rm3l) [reply ui/src/app/controls/endpoints/endpoints.component.html]: Done in [`d588446` (#7084)](https://github.com/redhat-developer/odo/pull/7084/commits/d5884465a3cb16b52e9851709f33625d96eeee14)

File: ui/src/app/controls/multi-key-value/multi-key-value.component.html (hunk 3 above)
Comment (feloy) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html]: Sorry, I was talking about the icons on the right of the Command / Args inputs (and probably others with the same disposition).

ui/src/app/controls/multi-text/multi-text.component.css:
```
+button.adjust-position {
+    right: 6px;
+    top: 6px;
+}
```

ui/src/app/controls/multi-text/multi-text.component.html:
```
-            <button [attr.data-cy]="dataCyPrefix+'-minus-'+i" mat-icon-button [matTooltip]="deleteLabel" (click)="removeText(i)">
+            <button class="adjust-position" [attr.data-cy]="dataCyPrefix+'-minus-'+i" mat-icon-button [matTooltip]="deleteLabel" (click)="removeText(i)">
```

Comment (rm3l) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html]: Ah right, I see. I'll make the changes. Thanks!

Comment (rm3l) [reply ui/src/app/controls/multi-key-value/multi-key-value.component.html]: Done in [`3001acb` (#7084)](https://github.com/redhat-developer/odo/pull/7084/commits/3001acb3c7ea61e584deafef375e9998751daa65)

**Review by feloy — APPROVED** [review]
Works great, thanks

**Comment by rm3l** [comment]
```
  [FAILED] Expected
      <string>:   __
       /  \__     Developing using the "lwvxbb" Devfile
       \__/  \    Namespace: cmd-logs-test249kbi
       /  \__/    odo version: v3.14.0 (b0412b0ee)
       \__/
      
      â†ª Running on the cluster in Dev mode
       â€¢  Waiting for Kubernetes resources  ...
       âš   Pod is Pending
       âœ“  Pod is Running
      --> Logs for lwvxbb-app-845986b854-nl689 / runtime
       â€¢  Syncing files into the container  ...
      
 âœ“  Syncing files into the container [457ms]
       â€¢  Building your application in container (command: install)  ...
      runtime: npm notice created a lockfile as package-lock.json. You should commit this file.
      runtime: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.2 (node_modules/chokidar/node_modules/fsevents):
      runtime: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
      runtime: npm WARN node-echo@0.0.1 No repository field.
      runtime: npm WARN node-echo@0.0.1 No license field.
      runtime: 
      runtime: added 33 packages from 29 contributors and audited 34 packages in 1.481s
      runtime: 
      runtime: 3 packages are looking for funding
      runtime:   run `npm fund` for details
      runtime: 
      runtime: found 0 vulnerabilities
      runtime: 
      
 âœ“  Building your application in container (command: install) [2s]
      
      â†ª Deploying Kubernetes Component: myapp-inner
       â€¢  Creating resource Pod/myapp-inner  ...
      
 âœ“  Creating resource Pod/myapp-inner 
       â€¢  Executing the application (command: start)  ...
      runtime: 
      runtime: > node-echo@0.0.1 start /projects
      runtime: > node server.js
      runtime: 
      runtime: Server running on 0.0.0.0:8080
       â€¢  Waiting for the application to be ready  ...
      
 âœ“  Waiting for the application to be ready [1s]
       -  Forwarding from 127.0.0.1:34059 -> 8080
      
      
      â†ª Dev mode
       Status:
       Watching for changes in the current directory /tmp/2912389663
      
      Keyboard Commands:
      [Ctrl+c] - Exit and delete resources from the cluster
           [p] - Manually apply local changes to the application on the cluster
      
  to contain substring
      <string>: main: 

[...]
Summarizing 1 Failure:
  [FAIL] odo logs command tests when component is created and odo logs is executed when running in Dev mode with --logs [It] 1. should successfully show logs of the running component
  /go/odo_1/tests/integration/cmd_logs_test.go:253

Ran 404 of 954 Specs in 1409.505 seconds
FAIL! -- 403 Passed | 1 Failed | 0 Pending | 550 Skipped
```

Flaky issue with `odo logs` - reported in #6619 (will reopen if the issue persists on other PRs)..

/override OpenShift-Integration-tests/OpenShift-Integration-tests

**Pull request merged**